pkg/frontend/querymiddleware/shard_active_series.go

```diff
@@ -0,0 +1,277 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/go-kit/log"
+	jsoniter "github.com/json-iterator/go"
+	"github.com/opentracing/opentracing-go"
+	otlog "github.com/opentracing/opentracing-go/log"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"golang.org/x/sync/errgroup"
+
+	apierror "github.com/grafana/mimir/pkg/api/error"
+	"github.com/grafana/mimir/pkg/distributor"
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/storage/sharding"
+	"github.com/grafana/mimir/pkg/util"
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+)
+
+type shardActiveSeriesMiddleware struct {
+	upstream http.RoundTripper
+	logger   log.Logger
+}
+
+func newShardActiveSeriesMiddleware(upstream http.RoundTripper, logger log.Logger) http.RoundTripper {
+	return &shardActiveSeriesMiddleware{
+		upstream: upstream,
+		logger:   logger,
+	}
+}
+
+func (s *shardActiveSeriesMiddleware) RoundTrip(r *http.Request) (*http.Response, error) {
+	const defaultNumShards = 1
+
+	ctx := r.Context()
+
+	spanLog, ctx := spanlogger.NewWithLogger(ctx, s.logger, "shardActiveSeries.RoundTrip")
+	defer spanLog.Finish()
+
+	numShards := setShardCountFromHeader(defaultNumShards, r, spanLog)
```

Comment:
Why don't we use the same values and logic as for usual query-sharding?

Reply:
I assume you mean using `getShardsForQuery` and `query_sharding_total_shards`?

Using a fixed shard count creates a lot of overhead and latency for cases where no or few shards are needed, which I think are the majority (I touched on this briefly in the PR description).

Ideally, we'd figure out how many series a selector touches and then shard the requests appropriately, but that's not easily possible at the moment 🤔

Reply:
> Using a fixed shard count creates a lot of overhead and latency for cases where no or few shards are needed, which I think are the majority

I don't understand why this is different from running an instant query `sum(...)` which would use `query_sharding_total_shards`: I know now we have the cardinality estimation cache, but previously we didn't have it and that wasn't an issue.

Reply:
I have changed the logic to default to a tenant's `query_sharding_total_shards` and allow this to be overridden via header.